### PR TITLE
Add notifications when there is an error in the service

### LINKED ISSUE #196: Add Notification when Accessing Pages such as Documents without Permission
**What would you like to be added**:
When users attempt to access pages or documents that they do not have permission to view, instead of simply being redirected or receiving an error page, it would be beneficial to add a notification to inform them of the reason why they cannot access the content. This will provide better user experience and clarity regarding access restrictions.

**Why is this needed**:
Enhancing the feedback provided to users when they encounter access restrictions will improve user experience and help users understand why they cannot view certain pages or documents.

**Additional Information**:

### PR DESCRIPTION
**Pull request by wet6123**
**What this PR does / why we need it**:

The currently implemented error exceptions don't have a detailed reason for the error, so users can't figure out why the error occurred. To solve this problem, I've added a detailed reason for each error exception and JWT AuthGuard. I've also modified the snackbar menu on the frontend to output the detailed reason.

Before modification
스크린샷 2024-07-25 오후 9 12 24   

After modification
스크린샷 2024-07-31 오후 8 52 06

**Which issue(s) this PR fixes**:

Fixes #196 

**Special notes for your reviewer**:

- You can check if the JWT token does not exist or expires.
- Added detailed error reasons to the exception that exists in the services.
- It would be nice to add error branches in the future to provide more detailed error handling.

**Does this PR introduce a user-facing change?**:

```release-note
Users can see why the error occurred.
```

**Additional documentation**:

```docs

```

**Checklist**:
- [x] Added relevant tests or not required
- [x] Didn't break anything

## Summary by CodeRabbit

- **New Features**
	- Enhanced JWT authentication with improved error handling and token verification.
	- Introduced a new `CustomError` interface in the frontend for better error handling.

- **Bug Fixes**
	- Improved clarity of error messages across multiple services, providing more detailed exceptions for unauthorized, not found, and conflict errors.
  
- **Documentation**
	- Updated error handling logic to prioritize specific feedback in the frontend.

### CODE REVIEW

File: backend/src/auth/jwt.guard.ts [hunk 1]
```diff
@@ -17,6 +27,38 @@ export class JwtAuthGuard extends AuthGuard("jwt") {
...
+				throw new UnauthorizedException("Unauthorized", {
+					cause: new Error(),
+					description: "Token not found",
+				});
+			}
+
+			const secretKey = this.configService.get<string>("JWT_AUTH_SECRET");
+			const decoded = jwt.verify(token, secretKey);
+			request.user = decoded;
+			return super.canActivate(context);
```
Comment (devleejb) [inline backend/src/auth/jwt.guard.ts:44]: We use jwt auth login provided from `AuthGuard`.
So, I don't want to implement jwt verify logic in CodePair.

Then, we can get the exception by overriding `handleRequest` in `AuthGuard`.

```typescript
handleRequest(err, user, info, context: ExecutionContext) {
    if (err || !user) {
	console.error("Authentication error:", err);
	console.error("Additional info:", info);

	// We can throw an exception we want
	throw err;
    }

    return user;
}
```

**Review by devleejb — CHANGES_REQUESTED** [review]
I have few questions.

**Q1)**
In constructor of nest js response error, I found that we can modify JSON response body by overriding `objectOrError`.

Should we use second parameter `descriptionOrOptions`?

```typescript
constructor(objectOrError?: string | object | any, descriptionOrOptions?: string | HttpExceptionOptions);
```

>   By default, the JSON response body contains two properties:
>   - `statusCode`: this will be the value 404.
>   - `message`: the string `'Not Found'` by default; override this by supplying  a string in the `objectOrError` parameter.

**Q2)**
It seems that the `cause` field in `descriptionOrOptions` are all `new Error()`. Should we pass the field?

File: backend/src/auth/jwt.guard.ts (hunk 1 above)
Comment (wet6123) [reply backend/src/auth/jwt.guard.ts]: I fixed it. Thank you!

**Comment by wet6123** [comment]
Thank you asking, @devleejb

**A1)**
I didn't want to lose the value of the *message*, so I didn't overwrite it. But I found that if we put the string at the `objectOrError`, the value of the *message* moved to the *error*, and the string value was saved in the *message*. I think this is better than the old way. Thanks for the advice!!

**A2)**
We don't have to. Totally my mistake. Thanks for letting me know!

**Review by devleejb — APPROVED** [review]
Thank you for your contribution.

**Pull request merged**